src/test/java/org/springframework/data/couchbase/transactions/CouchbaseTransactionManagerTransactionalTemplateIntegrationTests.java

```diff
@@ -0,0 +1,338 @@
+		AtomicInteger tryCount = new AtomicInteger(0);
+		UUID id = UUID.randomUUID();
+		Person person = new Person(id, "Walter", "White");
+		operations.insertById(Person.class).one(person);
+
+		personService.fetchAndRemove(person.getId().toString(), tryCount);
+
+		Person fetched = operations.findById(Person.class).one(person.getId().toString());
+		assertNull(fetched);
+		assertEquals(1, tryCount.get());
+	}
+
+	// gp: this test fails because it raises SimulateFailureException directly, rather than TransactionFailedException.
+	// This is because CouchbaseTransactionManager is not doing (and probably cannot do) the requisite error handling.
+	@DisplayName("Basic test of doing an insert then rolling back")
+	@Test
+	public void rollbackInsert() {
+		AtomicInteger tryCount = new AtomicInteger(0);
+		UUID id = UUID.randomUUID();
+
+		try {
+			personService.doInTransaction(tryCount, (ops) -> {
+				Person person = new Person(id, "Walter", "White");
+				ops.insertById(Person.class).one(person);
+				throw new SimulateFailureException();
+			});
+			fail();
+		} catch (TransactionFailedException err) {
+			assertTrue(err.getCause() instanceof SimulateFailureException);
+		}
+
+		Person fetched = operations.findById(Person.class).one(id.toString());
+		assertNull(fetched);
+		assertEquals(1, tryCount.get());
+	}
+
+	// gp: this test is failing for same reason as rollbackInsert
+	@DisplayName("Basic test of doing a replace then rolling back")
+	@Test
+	public void rollbackReplace() {
+		AtomicInteger tryCount = new AtomicInteger(0);
+		UUID id = UUID.randomUUID();
+		Person person = new Person(id, "Walter", "White");
+		operations.insertById(Person.class).one(person);
+
+		try {
+			personService.doInTransaction(tryCount, (ops) -> {
+				Person p = ops.findById(Person.class).one(person.getId().toString());
+				p.setFirstname("changed");
+				ops.replaceById(Person.class).one(p);
+				throw new SimulateFailureException();
+			});
+			fail();
+		} catch (TransactionFailedException err) {
+			assertTrue(err.getCause() instanceof SimulateFailureException);
+		}
+
+		Person fetched = operations.findById(Person.class).one(person.getId().toString());
+		assertEquals("Walter", fetched.getFirstname());
+		assertEquals(1, tryCount.get());
+	}
+
+	// gp: this test is failing for same reason as rollbackInsert
+	@DisplayName("Basic test of doing a remove then rolling back")
+	@Test
+	public void rollbackRemove() {
+		AtomicInteger tryCount = new AtomicInteger(0);
+		UUID id = UUID.randomUUID();
+		Person person = new Person(id, "Walter", "White");
+		operations.insertById(Person.class).one(person);
+
+		try {
+			personService.doInTransaction(tryCount, (ops) -> {
+				Person p = ops.findById(Person.class).one(person.getId().toString());
+				ops.removeById(Person.class).oneEntity(p);
+				throw new SimulateFailureException();
+			});
+			fail();
+		} catch (TransactionFailedException err) {
+			assertTrue(err.getCause() instanceof SimulateFailureException);
+		}
```

Comment:
There is an (inherited) JavaIntegrationTests.assertThrowsWithCause( runnable, exception... )  that does similar to the try/catch  [edit: I see you already found it].  (Actually the try/catch is easier to read when the runnable gets long).

Reply:
Yep I've been using that for new tests - I must have missed a couple.